File: 120. Triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        '''
        Modify original
        '''
        if not triangle:
            return 0
        for i in range(1, len(triangle)):
            for j in range(1,i):
                triangle[i][j] += min(triangle[i-1][j], triangle[i-1][j-1])
            triangle[i][0] += triangle[i-1][0]
            triangle[i][-1] += triangle[i-1][-1]
        return min(triangle[-1])

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        '''
        O(n) extra space
        '''
        if not triangle:
            return 0
        res = list(triangle[-1])
        for i in range(len(triangle)-2, -1, -1):
            for j in range(i+1):
                res[j] = min(res[j], res[j+1]) + triangle[i][j]
        return res[0]

File: 120. Triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        if (triangle==null || triangle.size()==0) {
            return 0;
        }
        int n = triangle.size();
        int[] res = new int[n];
        res[0] = triangle.get(0).get(0);
        for (int i=1; i<n; i++) {
            int[] temp = new int[n];
            List<Integer> level = triangle.get(i);
            for (int j=0; j<=i; j++) {
                if (j==0) {
                    temp[j] = res[j] + level.get(j);
                } else if (j==i) {
                    temp[j] = res[j-1] + level.get(j);
                } else {
                    temp[j] = Math.min(res[j-1], res[j]) + level.get(j);
                }
            }
            res = temp;
        }
        int minPath = Integer.MAX_VALUE;
        for (int i : res) {
            if (i < minPath) {
                minPath = i;
            }
        }
        return minPath;
    }
}